RecoLocalTracker/SiPixelRecHits/plugins/alpaka/SiPixelRecHitExtendedAlpaka.cc

```diff
@@ -0,0 +1,198 @@
+#include "HeterogeneousCore/AlpakaCore/interface/alpaka/Event.h"
+#include "HeterogeneousCore/AlpakaCore/interface/alpaka/EventSetup.h"
+#include "HeterogeneousCore/AlpakaCore/interface/alpaka/global/EDProducer.h"
+#include "HeterogeneousCore/AlpakaInterface/interface/config.h"
+#include "HeterogeneousCore/AlpakaInterface/interface/memory.h"
+#include "RecoLocalTracker/Records/interface/PixelCPEFastParamsRecord.h"
+#include "RecoLocalTracker/SiPixelRecHits/interface/PixelCPEBase.h"
+#include "RecoLocalTracker/SiPixelRecHits/interface/alpaka/PixelCPEFastParamsCollection.h"
+#include "RecoLocalTracker/SiPixelRecHits/interface/pixelCPEforDevice.h"
+
+#include "PixelRecHitKernel.h"
+
+#include <optional>
+
+//#define GPU_DEBUG
+
+namespace ALPAKA_ACCELERATOR_NAMESPACE {
+
+  class SiPixelRecHitExtendedAlpaka : public global::EDProducer<> {
+  public:
+    explicit SiPixelRecHitExtendedAlpaka(const edm::ParameterSet& iConfig);
+    ~SiPixelRecHitExtendedAlpaka() override = default;
+
+    static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+
+  private:
+    void produce(edm::StreamID streamID, device::Event& iEvent, const device::EventSetup& iSetup) const override;
+
+    const device::EDGetToken<reco::TrackingRecHitsSoACollection> pixelRecHitToken_;
+    const device::EDGetToken<reco::TrackingRecHitsSoACollection> trackerRecHitToken_;
+
+    const device::EDPutToken<reco::TrackingRecHitsSoACollection> outputRecHitsSoAToken_;
+  };
+
+  SiPixelRecHitExtendedAlpaka::SiPixelRecHitExtendedAlpaka(const edm::ParameterSet& iConfig)
+      : EDProducer(iConfig),
+        pixelRecHitToken_(consumes(iConfig.getParameter<edm::InputTag>("pixelRecHitsSoA"))),
+        trackerRecHitToken_(consumes(iConfig.getParameter<edm::InputTag>("trackerRecHitsSoA"))),
+        outputRecHitsSoAToken_(produces()) {}
+
+  void SiPixelRecHitExtendedAlpaka::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+    edm::ParameterSetDescription desc;
+
+    desc.add<edm::InputTag>("pixelRecHitsSoA", edm::InputTag("hltPhase2SiPixelRecHitsSoA"));
+    desc.add<edm::InputTag>("trackerRecHitsSoA", edm::InputTag("phase2OTRecHitsSoAConverter"));
+
+    descriptions.addWithDefaultLabel(desc);
+  }
+  // This utility unrolls the SoA columns (tuples) at compile time, calling the provided functor 'f'
+  // once for each element. The index is passed as a std::integral_constant so it
+  // is available at compile time.
+  template <std::size_t N, typename F, std::size_t... Is>
+  void unrollColumns(F&& f, std::index_sequence<Is...>) {
+    (f(std::integral_constant<std::size_t, Is>{}), ...);
+  }
+  // User-facing wrapper to deduce the size of the tuple and create the index sequence
+  // Usage: mergeSoAColumns<NumberOfColumns>([&](auto columnIndex) { ... });
+  template <std::size_t N, typename F>
+  void mergeSoAColumns(F&& f) {
+    unrollColumns<N>(std::forward<F>(f), std::make_index_sequence<N>{});
+  }
```

Comment:
Could you move these to an anonymous namespace, so they are only visible in this file ?